tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/Reduce.java

```diff
@@ -0,0 +1,239 @@
+  protected final MetricReduction reduction;
+  private final String totalName;
+  private final String countName;
+
+  private final Class<T> type;
+  /** the variable that holds the total of the metric values */
+  protected Variable<T> total;
+  /** the variable that holds the count of the metric values */
+  protected Variable<T> count;
+
+  /**
+   * Creates a Reducible Metric with a metric reductions of {@link MetricReduction#SUM}
+   *
+   * @param tf the TensorFlow Ops
+   * @param name the name for this metric. If null, name defaults to {@link Class#getSimpleName()}.
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and data type.
+   * @param type the type for the variables and result
+   */
+  protected Reduce(Ops tf, String name, long seed, Class<T> type) {
+    this(tf, name, MetricReduction.SUM, seed, type);
+  }
+
+  /**
+   * @param tf The TensorFlow Ops
+   * @param name the name for this metric. If null, name defaults to {@link Class#getSimpleName()}.
+   * @param reduction The type of metric reduction to apply
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and data type.
+   * @param type the type for the variables and result
+   */
+  protected Reduce(Ops tf, String name, MetricReduction reduction, long seed, Class<T> type) {
+    super(tf, name, seed);
+    this.reduction = reduction;
+    this.totalName = this.getVariableName(TOTAL);
+    this.countName = this.getVariableName(COUNT);
+    this.type = type;
+    setupVars();
+  }
+  /** Initializes the Variables */
+  private void setupVars() {
+    if (total == null) {
+      total = getTF().withName(totalName).variable(Shape.scalar(), type);
+    }
+    if (reduction == MetricReduction.SUM_OVER_BATCH_SIZE
+        || reduction == MetricReduction.WEIGHTED_MEAN) {
+      if (count == null) {
+        count = getTF().withName(countName).variable(Shape.scalar(), type);
+      }
+    }
+  }
+
+  /** {@inheritDoc} */
+  public Op resetStates() {
+    List<Op> controls = new ArrayList<>();
+    if (total != null) {
+      controls.add(
+          getTF().assign(total, CastHelper.cast(getTF(), getTF().constant(0), total.type())));
+    }
+    if (count != null) {
+      controls.add(
+          getTF().assign(count, CastHelper.cast(getTF(), getTF().constant(0), count.type())));
+    }
+    return getTF().withControlDependencies(controls).noOp();
+  }
+
+  /**
+   * Updates the metric variables based on the inputs. At least one input arg required for <code>
+   * values</code>, an optional additional input for the <code>sampleWeights</code>
+   *
+   * @param values the inputs to be passed to update state, this may not be null
+   * @param sampleWeights sample weights to be applied to values, may be null.
```

Comment:
The supported shape relationships between `values` and `sampleWeights` are complex. Do we want to document and test them?

Reply:
Broadcasting seems to be inherent throughout Numpy and TF Python. The most concise definition is from `tf.broadcast_to() `in TF Python. 

`Broadcasting is the process of making arrays to have compatible shapes for arithmetic operations. Two shapes are compatible if for each dimension pair they are either equal or one of them is one. When trying to broadcast a Tensor to a shape, it starts with the trailing dimensions, and works its way forward.`

If shapes aren't compatible, the low level TF Ops will throw TF Exceptions. These methods that test the shapes are trying to catch this before the TF Ops does as this makes it harder to debug the actual problem. 

Here is a link to `Numpy` broadcast rules. [Broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html).

It might be helpful to document this, but maybe in separate document like `Numpy` does.

Reply:
If it were simply that this method supports broadcasting, then I'd see no reason to say much about it at this stage of the project. But I think the permitted shape relationships are far more complex than that:

- We use `squeezeOrExpandDimensions` to "squeeze or expand the last dim of `sampleWeight` if its rank differs by 1 from [the rank of `values`]."

- Then we assert either at method runtime or at graph execution time (depending on our inputs) that "the `sampleWeights` can be broadcast to `values`." (I think we intentionally disallow the case where `values` would be broadcast to `sampleWeights`?)

- But if the inputs contain enough static shape information (how much is enough?) to allow us to check broadcasting at method runtime, and if we discover that the shapes don't support the unidirectional broadcasting that we allow, then instead we apply either `reduceSum` or `math.mean` (depending on `this.reduction`) to reduce `values` down to the same rank as `sampleWeights`.

- After all of the above logic, we make a broadcasting call to `math.mul`. So, for example, the previous step may use `reduceSum` to bring `values` to the same rank as `sampleWeights`, after which we may still broadcast one or more dimensions (having size 1) of `values` or `sampleWeights`, or both.

Reply:
I found the test cases for broadcasting in TF Python so I am creating Java test cases. I already found an anomaly with dynamic shapes and scalar, so I will investigate that, and when I'm done with the test cases I will check them in. 

Also, the method name was wrong and I have corrected it to `assertBroadcastable` to match TF Python.

Reply:
👍

Reply:
I finally took some time to study `tensorflow/python/keras/metrics.py`. I see that they apply  `squeeze_or_expand_dimensions` to `y_pred, y_true` rather than involving `sampleWeights`.

Reply:
`Reduce.java` line 131, `LossesHelper.squeezeOrExpandDimensions(getTF(), null, lValues, lSampleWeights);`
`null` is the `labels` parameter, which is `y_true` in the python version, I don't understand your point here.

I always map `y_true` to `labels`, and `y_pred` to `predictions`.

From `LossesHelper.java`
```
public static <T extends TNumber> LossTuple<T> squeezeOrExpandDimensions(
      Ops tf, Operand<T> labels, Operand<T> predictions, Operand<T> sampleWeights) 
```

from `metrics.py`
```
values, _, sample_weight = tf_losses_utils.squeeze_or_expand_dimensions(
          values, sample_weight=sample_weight)
```
Here `values` is `y_pred`.

from `python/ops/losses/util.py`
`def squeeze_or_expand_dimensions(y_pred, y_true=None, sample_weight=None):`

Reply:
Oops, sorry, I wasn't reading the code I thought I was! Your response is dead-on for the code I thought I was reading: `Reduce.update_state`.

Reply:
My concern that opened this thread remains open:

The supported shape relationships between values and sampleWeights are complex. Do we want to document and test them?

Reply:
I have added methods `assertBroadcastable` and `broadcastWeights` to `MetricsHelper` along with corresponding test classes. This should address the testing part, at least for these two topics. However, there are shape operations sprinkled all around the code, not just Metrics. I am not sure how feasible it is to capture all these. Much of the Python wrapper classes around low level operations involves manipulating shapes into shapes that are acceptable to the low level Op.

Reply:
Well, the Python code doesn't document the permitted shape relationships either. No reason this PR has to exceed the Python state of the art in this area!

👍 nothing to do